aeron-driver/src/main/c/aeron_driver_conductor.c

```diff
@@ -2789,13 +2817,198 @@ void aeron_driver_conductor_on_unavailable_image(
     aeron_mpsc_rb_t *receiver_rb = conductor->context->receiver_proxy->command_queue;
     return
         ((sender_rb->capacity - aeron_mpsc_rb_size(sender_rb)) <= AERON_COMMAND_RB_RESERVE) ||
         ((receiver_rb->capacity - aeron_mpsc_rb_size(receiver_rb)) <= AERON_COMMAND_RB_RESERVE);
 }
 
+typedef struct aeron_driver_async_command_stct
+{
+    aeron_udp_channel_async_parse_t async_parse;
+    aeron_name_resolver_async_resolve_t async_resolve;
+    aeron_send_channel_endpoint_t *endpoint;
+    aeron_uri_t *uri;
+    void *original_command;
+    bool is_exclusive;
+}
+aeron_driver_async_command_t;
+
+typedef int (*aeron_driver_async_client_command_on_execute_func_t)(aeron_driver_conductor_t *conductor, void *clientd);
+
+typedef int (*aeron_driver_async_client_command_on_complete_func_t)(aeron_driver_conductor_t *conductor, aeron_driver_async_command_t *async_command, void *clientd);
+
+typedef int (*aeron_driver_async_client_command_on_error_func_t)(aeron_driver_conductor_t *conductor, int result, aeron_driver_async_command_t *async_command, void *clientd);
+
+typedef struct aeron_driver_async_client_command_stct
+{
+    aeron_correlated_command_t *correlated;
+    aeron_driver_async_client_command_on_execute_func_t on_execute;
+    void *on_execute_clientd; // this is passed to the on_execute callback
+    aeron_driver_async_client_command_on_complete_func_t on_complete;
+    aeron_driver_async_client_command_on_error_func_t on_error;
+    aeron_driver_async_command_t async_command;
+}
+aeron_driver_async_client_command_t;
+
+/* This is an aeron_executor 'execute' callback - it's called from an executor thread */
+int aeron_driver_async_client_command_execute(void *task_clientd, void *executor_clientd)
+{
+    aeron_driver_async_client_command_t *async_client_command = task_clientd;
+    aeron_driver_conductor_t *conductor = executor_clientd;
+
+    if (async_client_command->on_execute(conductor, async_client_command->on_execute_clientd) < 0)
+    {
+        AERON_APPEND_ERR("%s", "");
+        return -1;
+    }
+
+    return 0;
+}
+
+/* This is an aeron_executor 'complete' callback - it's called when 'aeron_executor_process_completions' is called */
+void aeron_driver_async_client_command_complete(int result, int errcode, const char *errmsg, void *task_clientd, void *executor_clientd)
+{
+    aeron_driver_async_client_command_t *async_client_command = task_clientd;
+    aeron_driver_conductor_t *conductor = executor_clientd;
+    int64_t correlation_id = async_client_command->correlated->correlation_id;
+
+    conductor->async_client_command_in_flight = false;
+
+    if (result < 0)
+    {
+        if (async_client_command->on_error == NULL)
+        {
+            aeron_driver_conductor_on_error(conductor, errcode, errmsg, correlation_id);
+        }
+        else
+        {
+            // TODO set current errcode/errmsg to what was handed to this function???
```

Comment:
I think we do have to set those if we are going to use them below as they are thread local variables.  Can we put together a system test that will trigger this code path to see what would happen?

Reply:
Right now, the only time the `async_client_command->on_error` is used we don't check to see what the error was.  The original code just set `ss_family = AF_UNSPEC` and proceeded as normal.  The new code sets the on_error callback so that we have an opportunity to set AF_UNSPEC, but it then just calls the _complete() callback directly.  So I'm not sure we've got a good way to test this out at the moment - i.e. even if I set the errcode and errmsg thread local variables here, nobody ever reads them, so it won't matter.